src/Logger/Adapter/Raygun.php

```diff
@@ -0,0 +1,161 @@
+                'error' => [
+                    'className' => $log->getAction(),
+                    'message' => $log->getMessage()
+                ],
+                'tags' => $tagsArray,
+                'userCustomData' => $log->getExtra(),
+                'user' => [
+                    'isAnonymous' => empty($log->getUser()),
+                    'identifier' => empty($log->getUser()) ? null : $log->getUser()->getId(),
+                    'email' => empty($log->getUser()) ? null : $log->getUser()->getEmail(),
+                    'fullName' => empty($log->getUser()) ? null : $log->getUser()->getUsername(),
+                ],
+                'breadcrumbs' => $breadcrumbsArray
+            ]
+        ];
+
+        // init curl object
+        $ch = curl_init();
+
+        // define options
+        $optArray = array(
+            CURLOPT_URL => 'https://api.raygun.com/entries',
+            CURLOPT_RETURNTRANSFER => true,
+            CURLOPT_POST => true,
+            CURLOPT_POSTFIELDS => \json_encode($requestBody),
+            CURLOPT_HEADEROPT => CURLHEADER_UNIFIED,
+            CURLOPT_HTTPHEADER => array('Content-Type: application/json', 'X-ApiKey: ' . $this->apiKey)
+        );
+
+        // apply those options
+        curl_setopt_array($ch, $optArray);
+
+        // execute request and get response
+        $result = curl_exec($ch);
+
+        // also get the error and response code
+        $errors = curl_error($ch);
+        $response = curl_getinfo($ch, CURLINFO_HTTP_CODE);
+
+        curl_close($ch);
+
+        return $response;
+    }
+
+    /**
+     * Raygun constructor.
+     *
+     * @param string $configKey
+     */
+    public function __construct(string $configKey)
+    {
+        $this->apiKey = $configKey;
+    }
+
+    public function validateLog(Log $log): bool
+    {
+        // Support all log types (as tag)
+        switch ($log->getType()) {
+            case Log::TYPE_ERROR:
+            case Log::TYPE_WARNING:
+            case Log::TYPE_DEBUG:
+            case Log::TYPE_INFO:
+                break;
+            default:
+                throw new Exception("Supported log types for this adapter are: TYPE_ERROR, TYPE_WARNING, TYPE_DEBUG, TYPE_INFO");
+        }
+
+
+        // Supported breadcrumb types: debug, info, warning, error
+        foreach($log->getBreadcrumbs() as $breadcrumb) {
+            switch ($breadcrumb->getType()) {
+                case Log::TYPE_INFO:
+                case Log::TYPE_DEBUG:
+                case Log::TYPE_VERBOSE:
+                case Log::TYPE_ERROR:
+                case Log::TYPE_WARNING:
+                    break;
+                default:
+                    throw new Exception("Supported breadcrumb types for this adapter are: TYPE_INFO, TYPE_DEBUG, TYPE_VERBOSE, TYPE_ERROR, TYPE_WARNING");
+            }
+        }
+
+        // support all environment types (as tag)
+        switch ($log->getEnvironment()) {
```

Comment:
We can have the same logic as suggested for log types. We have a similar approach for db features in `utopia/database` for checking what abilities each adapter supports.

Reply:
Same as above applies. Checks can get specific in different providers and function `validateLog` gives the most flexibility.

Reply:
We could do it. Could look like
```
const supportedLogTypes = [
   TYPE_DEBUG,
   TYPE_ERROR,.  ....
]

getSupportForLogTypes (string $type) {
         return if supportedLogTypes contains $type
}
```

With type and environment, it might make sense. But what about `getSupportForBreadcrumbTypes()`? There might be adapter that does not support at all. It doesn't support breadcrumb, but it has to have a function that returns supported types? That's the dilemma we are facing - if we want functions that are not fully general.

Reply:
`getSupportForBreadcrumbTypes()` might work better if we change it to `getSupportForBreadcrumb(): bool`

Reply:
@eldadfux are you sure `getSupportForBreadcrumb()` and other specific functions would be a good idea to have in the abstract class ? As we mentioned before, not all providers would have concepts like breadcrumbs, logTypes, environmentTypes and so on. 

In the future, if another provider introduces a new concept like XYZ , we will need to add the `getSupportForXYZ()` to the abstract class and all other existing classes.  Do you think this would be a good pattern being too specific ? or we can keep it pretty open like `validate()`

Reply:
Implemented `getSupportedTypes`, `getSupportedEnvironments`, `getSupportedBreadcrumbTypes` in https://github.com/utopia-php/logger/commit/74140bf8bf9cb8aeb19f7505506e9cd9b71b725b